### PR TITLE
Use client-side websockets to call IBM Speech to Text

### PR DESCRIPTION
**Pull request by sandlerben**

### CODE REVIEW

**Comment by sandlerben** [comment]
r @yoninachmany @rachelhongca

File: transcription/ibm.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+// IBMResult is the result of an IBM transcription. See
```
Comment (yoninachmany) [inline transcription/ibm.go:16]: Nit: "For details, see" and delete "for details in line 17.

**Comment by yoninachmany** [comment]
reading through now -- what's the plan for testing?

**Comment by sandlerben** [comment]
I think I'll add mock testing when I get a chance.

On Wed, Apr 13, 2016 at 8:04 PM, Yoni Nachmany notifications@github.com
wrote:

> reading through now -- what's the plan for testing?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> https://github.com/hack4impact/transcribe4all/pull/39#issuecomment-209696262

## 

Benjamin Sandler
University of Pennsylvania | Class of 2018
Candidate for BS in Computer Science & BA in Cognitive Science
954.478.6689 | ben1sandler@gmail.com

File: transcription/ibm.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+		}
+		if len(result.Results) > 0 {
+			return result, nil
+		}
+	}
+}
+
+func basicAuth(username, password string) string {
+	auth := username + ":" + password
+	return base64.StdEncoding.EncodeToString([]byte(auth))
```
Comment (yoninachmany) [inline transcription/ibm.go:94]: What's the deal with `StdEncoding,` if both arguments and auth are already strings?

File: transcription/ibm.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	Final        bool                   `json:"final"`
+}
+type ibmAlternativesField struct {
+	WordConfidence    []ibmWordConfidence `json:"word_confidence"`
+	OverallConfidence float64             `json:"confidence"`
+	Transcript        string              `json:"transcript"`
+	Timestamps        []ibmWordTimestamp  `json:"timestamps"`
+}
+type ibmWordConfidence []interface{}
+type ibmWordTimestamp []interface{}
```
Comment (yoninachmany) [inline transcription/ibm.go]: Can you explain this type? Seems like a timestamp is defined as a list of three elements (described below):

> Time alignments for each word from the transcript as a list of lists. Each inner list consists of three elements: the word followed its start and end time in hundredths of seconds. For example, [["hello",0.0,1.2],["world",1.2,2.5]]. Available only for the best alternative.

File: transcription/ibm.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	Alternatives []ibmAlternativesField `json:"alternatives"`
+	Final        bool                   `json:"final"`
+}
+type ibmAlternativesField struct {
+	WordConfidence    []ibmWordConfidence `json:"word_confidence"`
+	OverallConfidence float64             `json:"confidence"`
+	Transcript        string              `json:"transcript"`
+	Timestamps        []ibmWordTimestamp  `json:"timestamps"`
+}
+type ibmWordConfidence []interface{}
```
Comment (yoninachmany) [inline transcription/ibm.go]: Can you explain this type? Seems like a confidence is well-defined list of two elements:

> A confidence score for each word of the transcript as a list of lists. Each inner list consists of two elements: the word and its confidence score in the range of 0 to 1. For example, [[\"hello\",0.95],[\"world\",0.866]]. Available only for the best alternative and only in results marked as final.

File: transcription/ibm.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+type ibmWordTimestamp []interface{}
+
+// TranscribeWithIBM transcribes a given audio file using the IBM Watson
+// Speech To Text API
+func TranscribeWithIBM(filePath string, IBMUsername string, IBMPassword string) (*IBMResult, error) {
+	result := new(IBMResult)
+
+	url := "wss://stream.watsonplatform.net/speech-to-text/api/v1/recognize?model=en-US_BroadbandModel"
```
Comment (yoninachmany) [inline transcription/ibm.go:41]: technically unnecessary to include model

> If you do not pass a model, the service uses the en-US_BroadbandModel by default.

File: transcription/ibm.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	header := http.Header{}
+	header.Set("Authorization", "Basic "+basicAuth(IBMUsername, IBMPassword))
+
+	dialer := websocket.DefaultDialer
+	ws, _, err := dialer.Dial(url, header)
```
Comment (yoninachmany) [inline transcription/ibm.go:46]: Any use for second return value (*http.Response) or does `err` do everything we need here?

File: transcription/ibm.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+	requestArgs := map[string]interface{}{
+		"action":             "start",
+		"content-type":       "audio/flac",
+		"continuous":         true,
+		"word_confidence":    true,
+		"timestamps":         true,
+		"profanity_filter":   false,
+		"interim_results":    false,
+		"inactivity_timeout": -1,
```
Comment (yoninachmany) [inline transcription/ibm.go:60]: Why should timeout be infinity?

File: transcription/ibm.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	header.Set("Authorization", "Basic "+basicAuth(IBMUsername, IBMPassword))
+
+	dialer := websocket.DefaultDialer
+	ws, _, err := dialer.Dial(url, header)
+	if err != nil {
+		return nil, err
+	}
+	defer ws.Close()
+
+	requestArgs := map[string]interface{}{
```
Comment (yoninachmany) [inline transcription/ibm.go:52]: Will `keywords` (and `keywords_threshold`) be included in next iteration?

File: transcription/ibm.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+func uploadFileWithWebsocket(ws *websocket.Conn, filePath string) error {
+	f, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+
+	r := bufio.NewReader(f)
+	buffer := make([]byte, 2048)
```
Comment (yoninachmany) [inline transcription/ibm.go:104]: Why 2048? Also is this being handled: "The WebSocket interface imposes a maximum frame size of 4 MB"?

File: transcription/ibm.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+		"interim_results":    false,
+		"inactivity_timeout": -1,
+	}
+	if err = ws.WriteJSON(requestArgs); err != nil {
+		return nil, err
+	}
+	if err = uploadFileWithWebsocket(ws, filePath); err != nil {
+		return nil, err
+	}
+	if err = ws.WriteMessage(websocket.BinaryMessage, []byte{}); err != nil { // write empty message to indicate end of uploading file
```
Comment (yoninachmany) [inline transcription/ibm.go]: Is it good style for the comment to be on the same line?

File: transcription/ibm.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+		return nil, err
+	}
+	log.Println("File uploaded")
+
+	// IBM must receive a message every 30 seconds or it will close the websocket.
```
Comment (yoninachmany) [inline transcription/ibm.go:74]: Why is this necessary if "inactivity_timeout": -1?

> The service terminates the session and closes the connection if the inactivity or session timeout is reached, as described in Timeouts. The inactivity timeout occurs if audio is being sent by the client but the service detects no speech. The inactivity timeout is 30 seconds by default; you can use the inactivity_timeout key to specify a different value.

**Comment by yoninachmany** [comment]
Added all my comments for now

File: transcription/ibm.go (hunk 4 above)
Comment (sandlerben) [inline transcription/ibm.go]: Yes, good question. 

From ["How to use interfaces in Go"](http://jordanorelli.com/post/32665860244/how-to-use-interfaces-in-go)

> The interface{} type, the empty interface is the interface that has no methods.

> Since there is no implements keyword, all types implement at least zero methods, and satisfying an interface is done automatically, all types satisfy the empty interface.
That means that if you write a function that takes an interface{} value as a parameter, you can supply that function with any value.

You will notice that the arrays returned by IBM contain both a string and a float64. Since typed slices in Go must contain all values of the **same** type, I created a slice of type interface{} since both strings and float64 satisfy the empty interface.

File: transcription/ibm.go (hunk 3 above)
Comment (sandlerben) [inline transcription/ibm.go]: Yes, good question. 

From ["How to use interfaces in Go"](http://jordanorelli.com/post/32665860244/how-to-use-interfaces-in-go)

> The interface{} type, the empty interface is the interface that has no methods.

> Since there is no implements keyword, all types implement at least zero methods, and satisfying an interface is done automatically, all types satisfy the empty interface.
That means that if you write a function that takes an interface{} value as a parameter, you can supply that function with any value.

You will notice that the arrays returned by IBM contain both a string and a float64. Since typed slices in Go must contain values of the **same** type, I created a slice of type interface{} since both strings and float64 satisfy the empty interface.

File: transcription/ibm.go (hunk 4 above)
Comment (sandlerben) [inline transcription/ibm.go]: Yes, good question. 

From ["How to use interfaces in Go"](http://jordanorelli.com/post/32665860244/how-to-use-interfaces-in-go)

> The interface{} type, the empty interface is the interface that has no methods.
> 
> Since there is no implements keyword, all types implement at least zero methods, and satisfying an interface is done automatically, all types satisfy the empty interface.
> That means that if you write a function that takes an interface{} value as a parameter, you can supply that function with any value.

You will notice that the arrays returned by IBM contain both a string and a float64. Since typed slices in Go must contain values of the **same** type, I created a slice of type interface{} since both strings and float64 satisfy the empty interface.

File: transcription/ibm.go (hunk 3 above)
Comment (sandlerben) [inline transcription/ibm.go]: See my comment above.

File: transcription/ibm.go (hunk 5 above)
Comment (sandlerben) [inline transcription/ibm.go:41]: Yes, this is true. I like to be explicit even when the default value is what I need. If the IBM library ever decides to change the default value, we will still be good.

File: transcription/ibm.go (hunk 6 above)
Comment (sandlerben) [inline transcription/ibm.go:46]: I don't think I need it.

File: transcription/ibm.go (hunk 8 above)
Comment (sandlerben) [inline transcription/ibm.go:52]: Yes!

File: transcription/ibm.go (hunk 7 above)
Comment (sandlerben) [inline transcription/ibm.go:60]: Basically, if there is x seconds of no speaking, we don't want IBM to end the session. If you look at IBM's own JavaScript demo, they also set the timeout to infinity.

File: transcription/ibm.go (hunk 10 above)
Comment (sandlerben) [inline transcription/ibm.go]: Hmm, I think this is our call but I think it would look better to move it up on its own line.

File: transcription/ibm.go (hunk 11 above)
Comment (sandlerben) [inline transcription/ibm.go:74]: The next paragraph says

> The session timeout occurs if the service receives no data from the client or sends no interim results for 30 seconds. You cannot change the length of this timeout. However, the client can extend the session by sending the service a JSON text message with the action key set to the value no-op: {"action": "no-op"}. This message touches the session and resets the timeout to keep the connection alive.

It's annoying and I don't quite understand why this is the behavior, but we do actually need to do this.

File: transcription/ibm.go (hunk 2 above)
Comment (sandlerben) [inline transcription/ibm.go:94]: I actually copied this function out of Go standard library's `http` package. I'm not entirely sure why the encoding is how it is, but you can read more about basic auth rules [here](https://en.wikipedia.org/wiki/Basic_access_authentication).

File: transcription/ibm.go (hunk 9 above)
Comment (sandlerben) [inline transcription/ibm.go:104]: That's a good question. I asked on the Gophers slack and this was the buffer size someone suggested. In theory, it shouldn't really make a difference.

**Comment by sandlerben** [comment]
@yoninachmany thanks for the great comments! I can tell you read my code and the IBM documentation closely 😄

File: transcription/ibm.go (hunk 9 above)
Comment (yoninachmany) [inline transcription/ibm.go:104]: Sounds good about 2048. How about this: "The WebSocket interface imposes a maximum frame size of 4 MB"?

Comment (sandlerben) [inline transcription/ibm.go:104]: Right. So the buffer could be much higher, but I don't see why we would want a higher buffer.

File: transcription/ibm.go (hunk 10 above)
Comment (yoninachmany) [inline transcription/ibm.go]: If so, please do!

File: transcription/ibm.go (hunk 1 above)
Comment (yoninachmany) [inline transcription/ibm.go:16]: Nit, but nice for comment to be 2 lines instead of 3.

**Comment by rachelhongca** [comment]
it looked good to me! (i didn't look at the code not yours)

File: transcription/ibm.go (hunk 2 above)
Comment (yoninachmany) [inline transcription/ibm.go:94]: lgtm!

File: transcription/ibm.go (hunk 5 above)
Comment (yoninachmany) [inline transcription/ibm.go:41]: lgtm!

File: transcription/ibm.go (hunk 6 above)
Comment (yoninachmany) [inline transcription/ibm.go:46]: lgtm!

File: transcription/ibm.go (hunk 7 above)
Comment (yoninachmany) [inline transcription/ibm.go:60]: lgtm!

File: transcription/ibm.go (hunk 8 above)
Comment (yoninachmany) [inline transcription/ibm.go:52]: lgtm!

File: transcription/ibm.go (hunk 9 above)
Comment (yoninachmany) [inline transcription/ibm.go:104]: lgtm!

File: transcription/ibm.go (hunk 11 above)
Comment (yoninachmany) [inline transcription/ibm.go:74]: lgtm!

**Comment by yoninachmany** [comment]
@sandlerben two minor nits about comments, but ready to merge!

File: transcription/ibm.go (hunk 1 above)
Comment (sandlerben) [inline transcription/ibm.go:16]: I was told to keep comments under 80 characters if possible. This is fine.
